validator_client/src/graffiti_file.rs

```diff
@@ -0,0 +1,166 @@
+use serde_derive::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{prelude::*, BufReader};
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use bls::blst_implementations::PublicKey;
+use types::{graffiti::GraffitiString, Graffiti};
+
+#[derive(Debug)]
+pub enum Error {
+    InvalidFile(std::io::Error),
+    InvalidLine,
+    InvalidPublicKey,
+    NoDefaultField,
+    InvalidGraffiti(String),
+}
+
+/// Struct to load validator graffitis from file.
+/// The graffiti file is expected to have the following structure
+///
+/// default: Lighthouse
+/// public_key1: graffiti1
+/// public_key2: graffiti2
+/// ...
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GraffitiFile {
+    graffiti_path: PathBuf,
+    graffitis: HashMap<PublicKey, Graffiti>,
+    default: Graffiti,
+}
+
+impl GraffitiFile {
+    pub fn new(graffiti_path: PathBuf) -> Self {
+        Self {
+            graffiti_path,
+            graffitis: HashMap::new(),
+            default: Default::default(),
+        }
+    }
+
+    /// Loads the graffiti file and populates the default graffiti and `graffitis` hashmap.
+    /// Returns the graffiti corresponding to the given public key if present, else returns the
+    /// default graffiti.
+    ///
+    /// Returns an error if loading from the graffiti file fails.
+    pub fn load_graffiti(&mut self, public_key: &PublicKey) -> Result<Graffiti, Error> {
+        self.read_graffiti_file()?;
+        Ok(self
+            .graffitis
+            .get(public_key)
+            .copied()
+            .unwrap_or(self.default))
+    }
+
+    /// Reads from a graffiti file with the specified format and populates the default value
+    /// and the hashmap.
+    ///
+    /// Returns an error if the file does not exist, or if the format is invalid.
+    pub fn read_graffiti_file(&mut self) -> Result<(), Error> {
+        let file = File::open(self.graffiti_path.as_path()).map_err(Error::InvalidFile)?;
+        let reader = BufReader::new(file);
+
+        let mut lines = reader.lines();
+
+        // Parse default
```

Comment:
AIUI, we require the default line to be the first line in the file. Would it be easier for users and code to allow it to appear anywhere in the file? If not, perhaps we could be a bit more clear about it in the docs.

Reply:
Allowed graffiti to appear anywhere.